### ISSUE TITLE
Support non-bubbling events with Actions Addon

### ISSUE THREAD

**Issue by rajsite** [storybookjs/storybook#17881]
**Is your feature request related to a problem? Please describe**

We have a web-component that fires and event named `cancel` following the API of the [native `dialog` element where `cancel` is not bubbling](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/cancel_event).

If we try to use:

```js
actions: {
    handles: ['cancel']
}
```
The action is never caught because non-bubbling events will not bubble up the DOM and event listeners need to be registered directly on the target element. The Actions addon assumes all events that are observed are bubbling events and assumes the event target is `document`:

https://github.com/storybookjs/storybook/blob/d377613e4274f57f875a926afd3931427d548982/addons/actions/src/preview/withActions.ts#L51

The **documentation is misleading** as it states that:

> This will bind a standard HTML event handler to the outermost HTML element rendered by your component

Where instead from the source it appears that the event listener is bound to `document`.

**Describe the solution you'd like**

I'd like to be able to listen for events that are not bubbling that are fired from my component. Some option ideas:

- One options is have a configurable root / target selector for each event that is configured with the Actions Addon. Maybe something like:

  ```js
  actions: {
    handles: [{
      name: 'cancel',
      rootSelector: 'my-element',
      eventTargetFilterSelector: null // potential name for the current selector configuration. Or it could continue to be part of name.
    }]
  }
  ```
- Another option is to change the target of the event listener to the expected event (seems like it might interfere with the current selector pattern).
- Another option might be to allow configuring the addEventListener mode to be capturing. Maybe as a per event setting:
  ```js
  actions: {
    handles: [{
      name: 'cancel',
      capture: true
    }]
  }
  ```

**Describe alternatives you've considered**

Couldn't find any workaround.

**Are you able to assist to bring the feature to reality?**

If there is a suggested API we may be able to help create a PR.

**Additional context**

N/A

**Comment by jhorback** [comment]
I found the same thing and agree with the suggestions. This is just basic web development 101 and I am surprised storybook doesn't support it. Forcing elements to bubble their events can cause unintended consequences. Event bubbling, especially when using a CustomEvent and web components, is a very specific design decision.

**Comment by RazeiXello** [comment]
+1 on this. Encountered yesterday when I was trying to figure out why my CustomEvents weren't being captured by Storybook Actions. Setting `bubbles` to `true` was the only way to see them.

**Comment by Benexx** [comment]
We do have the same issue when using webComponents. We use a lot of events and we don't want to bubble every event. Is there a workaround? Currently the only solution seems to create a custom render (with `StoryFn`) with every property and binding the events with `action('')`. That is a lot of (maintenance) work.

**Comment by frank-topel-dbi** [comment]
As the ticket author points out, what the documentation claims and what the actual actions plugin code does are two different things. 

Of course the actions plugin should be able to pick up on non-bubbling events! The dialog close event is just one example of a multitude of events that don't ever bubble. `focus` and `blur` are even more prominent examples (except those have bubbling pendants in the shape of `focusin` and `focusout`).

Would be willing to contribute to a solution, but none of the maintainers has even commented so far.

**Comment by frank-topel-dbi** [comment]
> We do have the same issue when using webComponents. We use a lot of events and we don't want to bubble every event. Is there a workaround? Currently the only solution seems to create a custom render (with `StoryFn`) with every property and binding the events with `action('')`. That is a lot of (maintenance) work.

Could you please lay out the workaround in a code example?

**Comment by pcvonz** [comment]
Yeah, just ran into this too. If nothing is done to fix this then at least the documentation should be updated...

But, this seems like a pretty easy change to make it work as the documentation claims it works:
https://github.com/storybookjs/storybook/blob/8b16feb67674d6ccfdf3a1c539ca35e56f3e1968/code/addons/actions/src/decorator.ts#L44

`const root = document && document.getElementById('storybook-root').children[0];`

🤔

**Comment by WilcoSp** [comment]
I've also ran into this and I've also tested both the [capture option](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#usecapture) and adding an event listener to each component/element.

Using the capture option would be the cleanest and simpler way to fix this issue:

I think the following way would be the best way to implement this while still being backwards compatible

```ts
interface ActionHandler {
    name: string // as "<event> <selector>"
    capture?: boolean
}


interface parameters {
    actions: {
        handles: (ActionHandler | string)[] // as a string it's still `<event> <selector>`, this is to be backwards compatible
        captureAll?: boolean //  can apply `capture` to all event listeners except set otherwise
    }
}

```

I would like to first know how you all think about this before I'll make a PR for this

**Comment by psyko-gh** [comment]
Any news/feedback on this ? Not being able to test/demo non bubbling events is too bad :(
